From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Joket_abc <15005771700@qq.com>
Date: Wed, 16 Nov 2022 22:37:53 +0800
Subject: [PATCH] Shears can have silk touch enchantment


diff --git a/src/main/java/org/bukkit/enchantments/EnchantmentTarget.java b/src/main/java/org/bukkit/enchantments/EnchantmentTarget.java
index 5eb81fcc18b8fdec5a0e4c699525281fa6ad4d78..5031673fc552c91419720cda9271a14136628f20 100644
--- a/src/main/java/org/bukkit/enchantments/EnchantmentTarget.java
+++ b/src/main/java/org/bukkit/enchantments/EnchantmentTarget.java
@@ -250,6 +250,18 @@ public enum EnchantmentTarget {
             return WEAPON.includes(item) || item.equals(Material.SHEARS);
         }
         // Purpur end
+        // KioCG start
+    },
+
+    /**
+     * Allow the Enchantment to be placed on shears.
+     */
+    TOOL_AND_SHEARS {
+        @Override
+        public boolean includes(@NotNull Material item) {
+            return TOOL.includes(item) || item.equals(Material.SHEARS);
+        }
+        // KioCG end
     };
 
     /**
