From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Joket_abc <15005771700@qq.com>
Date: Mon, 11 Dec 2023 20:09:31 +0800
Subject: [PATCH] Build changes


diff --git a/build.gradle.kts b/build.gradle.kts
index fb1837bd5d0cb1970a6c236e72654a1c09502c48..9b43246dce03177211862ff0291846f6bcdc6458 100644
--- a/build.gradle.kts
+++ b/build.gradle.kts
@@ -28,7 +28,7 @@ repositories {
 dependencies {
     extraRuntime(platform("net.kyori:adventure-bom:4.15.0-SNAPSHOT"))
     // Purpur start
-    implementation(project(":purpur-api"))
+    implementation(project(":kiocg-api")) // KioCG
     implementation("io.papermc.paper:paper-mojangapi:${project.version}") {
         exclude("io.papermc.paper", "paper-api")
     }
@@ -92,7 +92,7 @@ tasks.jar {
         attributes(
             "Main-Class" to "org.bukkit.craftbukkit.Main",
             "Implementation-Title" to "CraftBukkit",
-            "Implementation-Version" to "git-Purpur-$implementationVersion", // Purpur
+            "Implementation-Version" to "git-kiocg-$implementationVersion", // Purpur // KioCG
             "Implementation-Vendor" to date, // Paper
             "Specification-Title" to "Bukkit",
             "Specification-Version" to project.version,
diff --git a/src/main/java/org/bukkit/craftbukkit/util/Versioning.java b/src/main/java/org/bukkit/craftbukkit/util/Versioning.java
index fb87620c742ff7912f5e8ccd2a7930dd605576d9..bfb879097c4d90d4715c341d7202d72bbf4a4e3e 100644
--- a/src/main/java/org/bukkit/craftbukkit/util/Versioning.java
+++ b/src/main/java/org/bukkit/craftbukkit/util/Versioning.java
@@ -11,7 +11,7 @@ public final class Versioning {
     public static String getBukkitVersion() {
         String result = "Unknown-Version";
 
-        InputStream stream = Bukkit.class.getClassLoader().getResourceAsStream("META-INF/maven/org.purpurmc.purpur/purpur-api/pom.properties"); // Purpur
+        InputStream stream = Bukkit.class.getClassLoader().getResourceAsStream("META-INF/maven/com.kiocg.purpurfork/kiocg-api/pom.properties"); // Purpur // KioCG
         Properties properties = new Properties();
 
         if (stream != null) {
