From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Joket_abc <15005771700@qq.com>
Date: Tue, 6 Feb 2024 17:01:57 +0800
Subject: [PATCH] Override upstream patches


diff --git a/src/main/java/io/papermc/paper/commands/FeedbackForwardingSender.java b/src/main/java/io/papermc/paper/commands/FeedbackForwardingSender.java
index e3a5f1ec376319bdfda87fa27ae217bff3914292..26c6a95966a7751cc22e25092bd6cced5f641ea7 100644
--- a/src/main/java/io/papermc/paper/commands/FeedbackForwardingSender.java
+++ b/src/main/java/io/papermc/paper/commands/FeedbackForwardingSender.java
@@ -49,7 +49,7 @@ public final class FeedbackForwardingSender extends ServerCommandSender {
 
     @Override
     public String getName() {
-        return "FeedbackForwardingSender";
+        return io.papermc.paper.configuration.GlobalConfiguration.get().kiocgConfig.feedbackForwardingSenderName; // KioCG
     }
 
     @Override
diff --git a/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java b/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java
index a6f58b3457b7477015c5c6d969e7d83017dd3fa1..8de3c3aac56415dca3815d5b103c7501f6d33f7e 100644
--- a/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java
+++ b/src/main/java/io/papermc/paper/configuration/GlobalConfiguration.java
@@ -341,4 +341,12 @@ public class GlobalConfiguration extends ConfigurationPart {
         public boolean disableChorusPlantUpdates = false;
         public boolean disableMushroomBlockUpdates = false;
     }
+
+    // KioCG start
+    public KiocgConfig kiocgConfig;
+
+    public class KiocgConfig extends ConfigurationPart {
+        public String feedbackForwardingSenderName = "MOSS";
+    }
+    // KioCG end
 }
diff --git a/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java b/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java
index 385ca2c1022e0985550a5cc2bbff953f1aa33f5c..cf2fec7817c4077e56b4db7932e9df3404b72bd1 100644
--- a/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java
+++ b/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java
@@ -545,4 +545,12 @@ public class WorldConfiguration extends ConfigurationPart {
             VANILLA, EIGENCRAFT, ALTERNATE_CURRENT
         }
     }
+
+    // KioCG start
+    public KiocgConfig kiocgConfig;
+
+    public class KiocgConfig extends ConfigurationPart {
+
+    }
+    // KioCG end
 }
diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
index b1dd8d66a524254a270a725f5f7a46f28e13b749..34747338383d045c206f863ecf578a680f339148 100644
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -1215,7 +1215,7 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
                     this.recentTps[2] = tps5.getAverage();
                     this.recentTps[3] = tps15.getAverage();
                     // Purpur end
-                    lagging = recentTps[0] < org.purpurmc.purpur.PurpurConfig.laggingThreshold; // Purpur
+                    lagging = recentTps[1] < org.purpurmc.purpur.PurpurConfig.laggingThreshold; // Purpur // KioCG - 防止TPS瞬时波动
                     tickSection = currentTime;
                 }
                 // Paper end - further improve server tick loop
diff --git a/src/main/java/net/minecraft/server/level/ServerPlayer.java b/src/main/java/net/minecraft/server/level/ServerPlayer.java
index 36b7686fdfeaab13cb1be9610ddc76fc70f6f6e5..66ea5837dbfa9762d395f1bd754912587eeca38d 100644
--- a/src/main/java/net/minecraft/server/level/ServerPlayer.java
+++ b/src/main/java/net/minecraft/server/level/ServerPlayer.java
@@ -2312,7 +2312,7 @@ public class ServerPlayer extends Player {
             if (afk) {
                 getBukkitEntity().setPlayerListName(org.purpurmc.purpur.PurpurConfig.afkTabListPrefix + prefix + scoreboardName + suffix + org.purpurmc.purpur.PurpurConfig.afkTabListSuffix, true);
             } else {
-                getBukkitEntity().setPlayerListName(prefix + scoreboardName + suffix);
+                getBukkitEntity().setPlayerListName(prefix + scoreboardName + suffix, true); // KioCG - use minimessage
             }
         }
 
diff --git a/src/main/java/net/minecraft/world/entity/LivingEntity.java b/src/main/java/net/minecraft/world/entity/LivingEntity.java
index bfa69b2c37f07132f8b31c12d26f4ceb074901a5..98f18bca51abb05a8c9b4fcb55011d4dbf7e51fa 100644
--- a/src/main/java/net/minecraft/world/entity/LivingEntity.java
+++ b/src/main/java/net/minecraft/world/entity/LivingEntity.java
@@ -428,7 +428,12 @@ public abstract class LivingEntity extends Entity implements Attackable {
                         double d1 = this.level().getWorldBorder().getDamagePerBlock();
 
                         if (d1 > 0.0D) {
-                            if (level().purpurConfig.teleportIfOutsideBorder && this instanceof ServerPlayer serverPlayer) { serverPlayer.teleport(io.papermc.paper.util.MCUtil.toLocation(level(), ((ServerLevel) level()).getSharedSpawnPos())); return; } // Purpur
+                            // KioCG start
+                            if (level().purpurConfig.teleportIfOutsideBorder && this instanceof ServerPlayer serverPlayer) {
+                                serverPlayer.teleport(io.papermc.paper.util.MCUtil.toLocation(level(), ((ServerLevel) level()).getWorldBorder().clampToBounds(serverPlayer.getX(), serverPlayer.getY(), serverPlayer.getZ())).toHighestLocation());
+                                return;
+                            } // Purpur
+                            // KioCG end
                             this.hurt(this.damageSources().outOfBorder(), (float) Math.max(1, Mth.floor(-d0 * d1)));
                         }
                     }
diff --git a/src/main/java/net/minecraft/world/entity/animal/SnowGolem.java b/src/main/java/net/minecraft/world/entity/animal/SnowGolem.java
index b75d07f3af4addbb306ecb6baacf1607ef65fc25..37be573be745376ae722eb493a3fff2ed8441c43 100644
--- a/src/main/java/net/minecraft/world/entity/animal/SnowGolem.java
+++ b/src/main/java/net/minecraft/world/entity/animal/SnowGolem.java
@@ -237,7 +237,7 @@ public class SnowGolem extends AbstractGolem implements Shearable, RangedAttackM
     @Override
     // Purpur start
     public java.util.List<ItemStack> generateDefaultDrops(int looting) {
-        if (org.purpurmc.purpur.PurpurConfig.allowShearsLooting) {
+        if (org.purpurmc.purpur.PurpurConfig.allowShearsLooting && false) { // KioCG - fix dupe
             java.util.ArrayList<ItemStack> list = new java.util.ArrayList<>();
             for (int i = 0; i < 1 + looting; i++) {
                 list.add(new ItemStack(Items.CARVED_PUMPKIN));
diff --git a/src/main/java/net/minecraft/world/inventory/AnvilMenu.java b/src/main/java/net/minecraft/world/inventory/AnvilMenu.java
index 2913d69fcff4b6df68586146b7323cea33eba74b..cbb1fff06b62444277ba6589b5f6ec0f6ce96aae 100644
--- a/src/main/java/net/minecraft/world/inventory/AnvilMenu.java
+++ b/src/main/java/net/minecraft/world/inventory/AnvilMenu.java
@@ -324,6 +324,7 @@ public class AnvilMenu extends ItemCombinerMenu {
                                     name = name.replace("&" + match, "\u00a7" + match.toLowerCase(java.util.Locale.ROOT));
                                 }
                                 //name = name.replaceAll("(?i)&([0-9a-fr])", "\u00a7$1");
+                                name = name.replaceAll("(?i)&#([0-9a-f])([0-9a-f])([0-9a-f])([0-9a-f])([0-9a-f])([0-9a-f])", "\u00a7x\u00a7$1\u00a7$2\u00a7$3\u00a7$4\u00a7$5\u00a7$6"); // KioCG
                             }
                             if (player.hasPermission("purpur.anvil.format")) {
                                 java.util.regex.Matcher matcher = java.util.regex.Pattern.compile("(?i)&([k-or])").matcher(name);
diff --git a/src/main/java/net/minecraft/world/level/block/FarmBlock.java b/src/main/java/net/minecraft/world/level/block/FarmBlock.java
index 0c39126ce51439cce05747161aba43bce33a12d8..ba2305201707a1880a0f27014388645782255e39 100644
--- a/src/main/java/net/minecraft/world/level/block/FarmBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/FarmBlock.java
@@ -110,6 +110,24 @@ public class FarmBlock extends Block {
     @Override
     public void fallOn(Level world, BlockState state, BlockPos pos, Entity entity, float fallDistance) {
         super.fallOn(world, state, pos, entity, fallDistance); // CraftBukkit - moved here as game rules / events shouldn't affect fall damage.
+        // KioCG start - moved from below
+        // Purpur start
+        if (world.purpurConfig.farmlandTramplingDisabled) return;
+        if (world.purpurConfig.farmlandTramplingOnlyPlayers && !(entity instanceof Player)) return;
+        if (world.purpurConfig.farmlandAlpha) {
+            Block block = world.getBlockState(pos.below()).getBlock();
+            if (block instanceof FenceBlock || block instanceof WallBlock) {
+                return;
+            }
+        }
+        if (world.purpurConfig.farmlandTramplingFeatherFalling) {
+            Iterator<net.minecraft.world.item.ItemStack> armor = entity.getArmorSlots().iterator();
+            if (armor.hasNext() && net.minecraft.world.item.enchantment.EnchantmentHelper.getItemEnchantmentLevel(net.minecraft.world.item.enchantment.Enchantments.FALL_PROTECTION, armor.next()) >= (int) entity.fallDistance) {
+                return;
+            }
+        }
+        // Purpur end
+        // KioCG end - moved from below
         if (!world.isClientSide && (world.purpurConfig.farmlandTrampleHeight >= 0D ? fallDistance >= world.purpurConfig.farmlandTrampleHeight : world.random.nextFloat() < fallDistance - 0.5F) && entity instanceof LivingEntity && (entity instanceof Player || world.purpurConfig.farmlandBypassMobGriefing || world.getGameRules().getBoolean(GameRules.RULE_MOBGRIEFING)) && entity.getBbWidth() * entity.getBbWidth() * entity.getBbHeight() > 0.512F) { // Purpur
             // CraftBukkit start - Interact soil
             org.bukkit.event.Cancellable cancellable;
@@ -124,22 +142,24 @@ public class FarmBlock extends Block {
                 return;
             }
 
-            // Purpur start
-            if (world.purpurConfig.farmlandTramplingDisabled) return;
-            if (world.purpurConfig.farmlandTramplingOnlyPlayers && !(entity instanceof Player)) return;
-            if (world.purpurConfig.farmlandAlpha) {
-                Block block = world.getBlockState(pos.below()).getBlock();
-                if (block instanceof FenceBlock || block instanceof WallBlock) {
-                    return;
-                }
-            }
-            if (world.purpurConfig.farmlandTramplingFeatherFalling) {
-                Iterator<net.minecraft.world.item.ItemStack> armor = entity.getArmorSlots().iterator();
-                if (armor.hasNext() && net.minecraft.world.item.enchantment.EnchantmentHelper.getItemEnchantmentLevel(net.minecraft.world.item.enchantment.Enchantments.FALL_PROTECTION, armor.next()) >= (int) entity.fallDistance) {
-                    return;
-                }
-            }
-            // Purpur end
+            // KioCG start
+//            // Purpur start
+//            if (world.purpurConfig.farmlandTramplingDisabled) return;
+//            if (world.purpurConfig.farmlandTramplingOnlyPlayers && !(entity instanceof Player)) return;
+//            if (world.purpurConfig.farmlandAlpha) {
+//                Block block = world.getBlockState(pos.below()).getBlock();
+//                if (block instanceof FenceBlock || block instanceof WallBlock) {
+//                    return;
+//                }
+//            }
+//            if (world.purpurConfig.farmlandTramplingFeatherFalling) {
+//                Iterator<net.minecraft.world.item.ItemStack> armor = entity.getArmorSlots().iterator();
+//                if (armor.hasNext() && net.minecraft.world.item.enchantment.EnchantmentHelper.getItemEnchantmentLevel(net.minecraft.world.item.enchantment.Enchantments.FALL_PROTECTION, armor.next()) >= (int) entity.fallDistance) {
+//                    return;
+//                }
+//            }
+//            // Purpur end
+            // KioCG end
             if (!CraftEventFactory.callEntityChangeBlockEvent(entity, pos, Blocks.DIRT.defaultBlockState())) {
                 return;
             }
diff --git a/src/main/java/org/purpurmc/purpur/entity/PhantomFlames.java b/src/main/java/org/purpurmc/purpur/entity/PhantomFlames.java
index 11825590af9346c61d5d15e5ef446b3c77b81b54..767dc9667872da828077e75b5871a789d3d1af27 100644
--- a/src/main/java/org/purpurmc/purpur/entity/PhantomFlames.java
+++ b/src/main/java/org/purpurmc/purpur/entity/PhantomFlames.java
@@ -98,7 +98,7 @@ public class PhantomFlames extends LlamaSpit {
         Entity shooter = this.getOwner();
         if (shooter instanceof LivingEntity) {
             Entity target = entityHitResult.getEntity();
-            if (canGrief || (target instanceof LivingEntity && !(target instanceof ArmorStand))) {
+            if (canGrief || target instanceof net.minecraft.world.entity.player.Player) { // KioCG - 只伤害玩家
                 boolean hurt = target.hurt(target.damageSources().mobProjectile(this, (LivingEntity) shooter), level().purpurConfig.phantomFlameDamage);
                 if (hurt && level().purpurConfig.phantomFlameFireTime > 0) {
                     target.setSecondsOnFire(level().purpurConfig.phantomFlameFireTime);
