From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Joket_abc <15005771700@qq.com>
Date: Tue, 17 Oct 2023 19:16:08 +0800
Subject: [PATCH] Lobotomized animals


diff --git a/src/main/java/com/kiocg/entity/Lobotomized.java b/src/main/java/com/kiocg/entity/Lobotomized.java
new file mode 100644
index 0000000000000000000000000000000000000000..4fe89ac58ed27baeac4105c1b4bfa163dd756a59
--- /dev/null
+++ b/src/main/java/com/kiocg/entity/Lobotomized.java
@@ -0,0 +1,51 @@
+package com.kiocg.entity;
+
+import net.minecraft.core.BlockPos;
+import net.minecraft.world.entity.LivingEntity;
+import net.minecraft.world.level.Level;
+import net.minecraft.world.level.pathfinder.PathComputationType;
+
+public class Lobotomized {
+    private boolean isLobotomized = false;
+
+    public boolean isLobotomized() {
+        return this.isLobotomized;
+    }
+
+    public void setLobotomized(boolean b) {
+        this.isLobotomized = b;
+    }
+
+    public boolean checkLobotomized(LivingEntity livingEntity) {
+        if (!livingEntity.onGround()) {
+            return this.isLobotomized = false;
+        }
+        int interval = livingEntity.level().purpurConfig.villagerLobotomizeCheckInterval;
+        if ((livingEntity.getId() + livingEntity.tickCount) % interval == 0) {
+            // offset Y for short blocks like dirt_path/farmland
+            // 扩展到半砖高度 (希望不会出问题!)
+            this.isLobotomized = !canTravelFrom(BlockPos.containing(livingEntity.getX(), livingEntity.getBoundingBox().minY + 0.5000001D, livingEntity.getZ()), livingEntity.level(), livingEntity.getBbHeight() <= 1.0F);
+        }
+        return this.isLobotomized;
+    }
+
+    private boolean canTravelFrom(BlockPos pos, Level level, boolean ignoreTop) {
+        return canTravelTo(pos.east(), level, ignoreTop) || canTravelTo(pos.west(), level, ignoreTop) || canTravelTo(pos.north(), level, ignoreTop) || canTravelTo(pos.south(), level, ignoreTop);
+    }
+
+    private boolean canTravelTo(BlockPos pos, Level level, boolean ignoreTop) {
+        net.minecraft.world.level.block.state.BlockState state = level.getBlockStateIfLoaded(pos);
+        if (state == null) {
+            // chunk not loaded
+            return false;
+        }
+
+        // 优化单格生物的检测
+        if (ignoreTop) {
+            return state.isPathfindable(level, pos, PathComputationType.LAND);
+        }
+
+        // only if both blocks have no collision
+        return state.isPathfindable(level, pos, PathComputationType.LAND) && level.getBlockState(pos.above()).isPathfindable(level, pos, PathComputationType.LAND);
+    }
+}
diff --git a/src/main/java/net/minecraft/world/entity/LivingEntity.java b/src/main/java/net/minecraft/world/entity/LivingEntity.java
index 28ab90dfbb6436e0180f8074fb7abf2a25532440..de3e3ce532b0b5c803273e122b4942d6dd41392e 100644
--- a/src/main/java/net/minecraft/world/entity/LivingEntity.java
+++ b/src/main/java/net/minecraft/world/entity/LivingEntity.java
@@ -305,6 +305,8 @@ public abstract class LivingEntity extends Entity implements Attackable {
         this.brain = this.makeBrain(new Dynamic(dynamicopsnbt, (Tag) dynamicopsnbt.createMap((Map) ImmutableMap.of(dynamicopsnbt.createString("memories"), (Tag) dynamicopsnbt.emptyMap()))));
     }
 
+    public boolean isLobotomized() { return false; } // KioCG
+
     protected void initAttributes() {}// Purpur
 
     public Brain<?> getBrain() {
diff --git a/src/main/java/net/minecraft/world/entity/ai/goal/FloatGoal.java b/src/main/java/net/minecraft/world/entity/ai/goal/FloatGoal.java
index 7093c62be53fe99ed9880fc8ddaa07440fe4f715..577637c3e9d3eae5639129a34a7d4d90fb00ad48 100644
--- a/src/main/java/net/minecraft/world/entity/ai/goal/FloatGoal.java
+++ b/src/main/java/net/minecraft/world/entity/ai/goal/FloatGoal.java
@@ -16,6 +16,7 @@ public class FloatGoal extends Goal {
 
     @Override
     public boolean canUse() {
+        if (this.mob.isLobotomized()) return false; // KioCG
         return this.mob.isInWater() && this.mob.getFluidHeight(FluidTags.WATER) > this.mob.getFluidJumpThreshold() || this.mob.isInLava();
     }
 
diff --git a/src/main/java/net/minecraft/world/entity/ai/goal/FollowParentGoal.java b/src/main/java/net/minecraft/world/entity/ai/goal/FollowParentGoal.java
index d8a2c1821fb0ec891c2f7b4e75d96a5403b4ddd1..30f0e034ae3098ad2fd4f35daac2a612120ac58c 100644
--- a/src/main/java/net/minecraft/world/entity/ai/goal/FollowParentGoal.java
+++ b/src/main/java/net/minecraft/world/entity/ai/goal/FollowParentGoal.java
@@ -21,6 +21,7 @@ public class FollowParentGoal extends Goal {
 
     @Override
     public boolean canUse() {
+        if (this.animal.isLobotomized()) return false; // KioCG
         if (this.animal.getAge() >= 0) {
             return false;
         } else {
@@ -51,6 +52,7 @@ public class FollowParentGoal extends Goal {
 
     @Override
     public boolean canContinueToUse() {
+        if (this.animal.isLobotomized()) return false; // KioCG
         if (this.animal.getAge() >= 0) {
             return false;
         } else if (!this.parent.isAlive()) {
diff --git a/src/main/java/net/minecraft/world/entity/ai/goal/LookAtPlayerGoal.java b/src/main/java/net/minecraft/world/entity/ai/goal/LookAtPlayerGoal.java
index dc1fb94161b32a103337dccdd89977086a106a46..c8fc20fcdf7cbd0936d5177b60a2a8d3b5b9cb14 100644
--- a/src/main/java/net/minecraft/world/entity/ai/goal/LookAtPlayerGoal.java
+++ b/src/main/java/net/minecraft/world/entity/ai/goal/LookAtPlayerGoal.java
@@ -48,6 +48,7 @@ public class LookAtPlayerGoal extends Goal {
 
     @Override
     public boolean canUse() {
+        if (this.mob.isLobotomized()) return false; // KioCG
         if (this.mob.getRandom().nextFloat() >= this.probability) {
             return false;
         } else {
@@ -69,6 +70,7 @@ public class LookAtPlayerGoal extends Goal {
 
     @Override
     public boolean canContinueToUse() {
+        if (this.mob.isLobotomized()) return false; // KioCG
         if (!this.lookAt.isAlive()) {
             return false;
         } else if (this.mob.distanceToSqr(this.lookAt) > (double)(this.lookDistance * this.lookDistance)) {
diff --git a/src/main/java/net/minecraft/world/entity/ai/goal/PanicGoal.java b/src/main/java/net/minecraft/world/entity/ai/goal/PanicGoal.java
index 3263c816e82285ff943270be21049432ac7f55e9..71cfe89bebf60a904607a21fbc0048e7403b2a2e 100644
--- a/src/main/java/net/minecraft/world/entity/ai/goal/PanicGoal.java
+++ b/src/main/java/net/minecraft/world/entity/ai/goal/PanicGoal.java
@@ -27,6 +27,7 @@ public class PanicGoal extends Goal {
 
     @Override
     public boolean canUse() {
+        if (this.mob.isLobotomized()) return false; // KioCG
         if (!this.shouldPanic()) {
             return false;
         } else {
@@ -77,6 +78,7 @@ public class PanicGoal extends Goal {
 
     @Override
     public boolean canContinueToUse() {
+        if (this.mob.isLobotomized()) return false; // KioCG
         return !this.mob.getNavigation().isDone();
     }
 
diff --git a/src/main/java/net/minecraft/world/entity/ai/goal/RandomStrollGoal.java b/src/main/java/net/minecraft/world/entity/ai/goal/RandomStrollGoal.java
index 9c57a2e4b1a80711efbf2a756458c0250a43f880..dac232e2e19ff43e5a6a1e6252b1555a489c5caf 100644
--- a/src/main/java/net/minecraft/world/entity/ai/goal/RandomStrollGoal.java
+++ b/src/main/java/net/minecraft/world/entity/ai/goal/RandomStrollGoal.java
@@ -35,6 +35,7 @@ public class RandomStrollGoal extends Goal {
 
     @Override
     public boolean canUse() {
+        if (this.mob.isLobotomized()) return false; // KioCG
         if (this.mob.hasControllingPassenger()) {
             return false;
         } else {
@@ -68,6 +69,7 @@ public class RandomStrollGoal extends Goal {
 
     @Override
     public boolean canContinueToUse() {
+        if (this.mob.isLobotomized()) return false; // KioCG
         return !this.mob.getNavigation().isDone() && !this.mob.hasControllingPassenger();
     }
 
diff --git a/src/main/java/net/minecraft/world/entity/ai/goal/TemptGoal.java b/src/main/java/net/minecraft/world/entity/ai/goal/TemptGoal.java
index 00cf59524477ec79d4354cc403fc3e75a63b81a0..a1cc5c1524208835f9eb2cf83802415544f94ab0 100644
--- a/src/main/java/net/minecraft/world/entity/ai/goal/TemptGoal.java
+++ b/src/main/java/net/minecraft/world/entity/ai/goal/TemptGoal.java
@@ -43,6 +43,7 @@ public class TemptGoal extends Goal {
 
     @Override
     public boolean canUse() {
+        if (this.mob.isLobotomized()) return false; // KioCG
         if (this.calmDown > 0) {
             --this.calmDown;
             return false;
@@ -67,6 +68,7 @@ public class TemptGoal extends Goal {
 
     @Override
     public boolean canContinueToUse() {
+        if (this.mob.isLobotomized()) return false; // KioCG
         if (this.canScare()) {
             if (this.mob.distanceToSqr((Entity) this.player) < 36.0D) {
                 if (this.player.distanceToSqr(this.px, this.py, this.pz) > 0.010000000000000002D) {
diff --git a/src/main/java/net/minecraft/world/entity/ai/navigation/PathNavigation.java b/src/main/java/net/minecraft/world/entity/ai/navigation/PathNavigation.java
index 0e36e3e672de4c806e2d534b437489a3a349a722..d701c94c55339c31f3a7ce82da5c714cc97a612e 100644
--- a/src/main/java/net/minecraft/world/entity/ai/navigation/PathNavigation.java
+++ b/src/main/java/net/minecraft/world/entity/ai/navigation/PathNavigation.java
@@ -198,6 +198,7 @@ public abstract class PathNavigation {
     // Paper end
 
     public boolean moveTo(Entity entity, double speed) {
+        if (this.mob.isLobotomized()) return false; // KioCG
         // Paper start - Pathfinding optimizations
         if (this.pathfindFailures > 10 && this.path == null && net.minecraft.server.MinecraftServer.currentTick < this.lastFailure + 40) {
             return false;
@@ -218,6 +219,7 @@ public abstract class PathNavigation {
     }
 
     public boolean moveTo(@Nullable Path path, double speed) {
+        if (this.mob.isLobotomized()) return false; // KioCG
         if (path == null) {
             this.path = null;
             return false;
diff --git a/src/main/java/net/minecraft/world/entity/animal/Animal.java b/src/main/java/net/minecraft/world/entity/animal/Animal.java
index a434d91b8dfff30cff81df964ea8149caa8cb604..ac5d8f4ae71acdeeae5901df2a70d562d2e242c1 100644
--- a/src/main/java/net/minecraft/world/entity/animal/Animal.java
+++ b/src/main/java/net/minecraft/world/entity/animal/Animal.java
@@ -42,6 +42,7 @@ public abstract class Animal extends AgeableMob {
     @Nullable
     public UUID loveCause;
     public ItemStack breedItem; // CraftBukkit - Add breedItem variable
+    protected com.kiocg.entity.Lobotomized lobotomized = new com.kiocg.entity.Lobotomized();
     public abstract int getPurpurBreedTime(); // Purpur
 
     protected Animal(EntityType<? extends Animal> type, Level world) {
@@ -77,8 +78,31 @@ public abstract class Animal extends AgeableMob {
             }
         }
 
+        this.lobotomized.checkLobotomized(this); // KioCG
     }
 
+    // KioCG start
+    @Override
+    public void travel(net.minecraft.world.phys.Vec3 vec3) {
+        if (!this.isLobotomized() || !this.passengers.isEmpty() || this.lastHurtByPlayer != null || this.lastHurtByMob != null || (this.tickCount + this.getId()) % 100 == 0) {
+            super.travel(vec3);
+        }
+    }
+
+    @Override
+    public boolean isCollidable(boolean ignoreClimbing) {
+        if (!this.isLobotomized() || !this.passengers.isEmpty() || this.lastHurtByPlayer != null || this.lastHurtByMob != null || net.minecraft.server.MinecraftServer.currentTick % 100 < 4) {
+            return super.isCollidable(ignoreClimbing);
+        }
+        return false;
+    }
+
+    @Override
+    public boolean isLobotomized() {
+        return this.lobotomized.isLobotomized();
+    }
+    // KioCG end
+
     @Override
     public boolean hurt(DamageSource source, float amount) {
         if (this.isInvulnerableTo(source)) {
@@ -106,6 +130,7 @@ public abstract class Animal extends AgeableMob {
         if (this.loveCause != null) {
             nbt.putUUID("LoveCause", this.loveCause);
         }
+        nbt.putBoolean("KioCG.Lobotomized", this.lobotomized.isLobotomized()); // KioCG
 
     }
 
@@ -114,6 +139,13 @@ public abstract class Animal extends AgeableMob {
         super.readAdditionalSaveData(nbt);
         this.inLove = nbt.getInt("InLove");
         this.loveCause = nbt.hasUUID("LoveCause") ? nbt.getUUID("LoveCause") : null;
+        // KioCG start
+        boolean isLobotomized = nbt.getBoolean("KioCG.Lobotomized");
+        this.lobotomized.setLobotomized(isLobotomized);
+        if (isLobotomized) {
+            this.onGround = true;
+        }
+        // KioCG end
     }
 
     public static boolean checkAnimalSpawnRules(EntityType<? extends Animal> type, LevelAccessor world, MobSpawnType spawnReason, BlockPos pos, RandomSource random) {
diff --git a/src/main/java/net/minecraft/world/entity/animal/Bee.java b/src/main/java/net/minecraft/world/entity/animal/Bee.java
index 38a3dcec138d9233a46e5d523bcc6d64bc7fffd0..5ebd4cb2ecdec3e12b7ee27e31e5bc324fa82825 100644
--- a/src/main/java/net/minecraft/world/entity/animal/Bee.java
+++ b/src/main/java/net/minecraft/world/entity/animal/Bee.java
@@ -1111,6 +1111,7 @@ public class Bee extends Animal implements NeutralMob, FlyingAnimal {
 
         @Override
         public boolean canBeeUse() {
+            if (Bee.this.isLobotomized()) return false; // KioCG
             return Bee.this.hivePos != null && !Bee.this.hasRestriction() && Bee.this.wantsToEnterHive() && !this.hasReachedTarget(Bee.this.hivePos) && Bee.this.level().getBlockState(Bee.this.hivePos).is(BlockTags.BEEHIVES);
         }
 
@@ -1227,6 +1228,7 @@ public class Bee extends Animal implements NeutralMob, FlyingAnimal {
 
         @Override
         public boolean canBeeUse() {
+            if (Bee.this.isLobotomized()) return false; // KioCG
             return Bee.this.savedFlowerPos != null && !Bee.this.hasRestriction() && this.wantsToGoToKnownFlower() && Bee.this.isFlowerValid(Bee.this.savedFlowerPos) && !Bee.this.closerThan(Bee.this.savedFlowerPos, 2);
         }
 
@@ -1279,6 +1281,7 @@ public class Bee extends Animal implements NeutralMob, FlyingAnimal {
 
         @Override
         public boolean canBeeUse() {
+            if (Bee.this.isLobotomized()) return false; // KioCG
             return Bee.this.getCropsGrownSincePollination() >= 10 ? false : (Bee.this.random.nextFloat() < 0.3F ? false : Bee.this.hasNectar() && Bee.this.isHiveValid());
         }
 
@@ -1343,6 +1346,7 @@ public class Bee extends Animal implements NeutralMob, FlyingAnimal {
 
         @Override
         public boolean canUse() {
+            if (Bee.this.isLobotomized()) return false; // KioCG
             return Bee.this.navigation.isDone() && Bee.this.random.nextInt(10) == 0;
         }
 
diff --git a/src/main/java/net/minecraft/world/entity/npc/Villager.java b/src/main/java/net/minecraft/world/entity/npc/Villager.java
index 02df0bf09f2ef91e4dafbdcea590b4634b2abf08..7a5409ae7f7dfea779564e94ffe84c3bec21e04c 100644
--- a/src/main/java/net/minecraft/world/entity/npc/Villager.java
+++ b/src/main/java/net/minecraft/world/entity/npc/Villager.java
@@ -141,7 +141,8 @@ public class Villager extends AbstractVillager implements ReputationEventHandler
     }, MemoryModuleType.MEETING_POINT, (entityvillager, holder) -> {
         return holder.is(PoiTypes.MEETING);
     });
-    private boolean isLobotomized = false; public boolean isLobotomized() { return this.isLobotomized; } // Purpur
+    private com.kiocg.entity.Lobotomized lobotomized = new com.kiocg.entity.Lobotomized(); // KioCG
+    private boolean isLobotomized = false; @Override public boolean isLobotomized() { return this.lobotomized.isLobotomized(); } // Purpur // KioCG - 使用更好的检测
     private int notLobotomizedCount = 0; // Purpur
 
     public long nextGolemPanic = -1; // Pufferfish
@@ -243,6 +244,23 @@ public class Villager extends AbstractVillager implements ReputationEventHandler
         return !bottom.hasCollision && !top.hasCollision;
     }
 
+    // KioCG start
+    @Override
+    public void travel(net.minecraft.world.phys.Vec3 vec3) {
+        if (!this.isLobotomized() || !this.passengers.isEmpty() || this.lastHurtByPlayer != null || this.lastHurtByMob != null || (this.tickCount + this.getId()) % 100 == 0) {
+            super.travel(vec3);
+        }
+    }
+
+    @Override
+    public boolean isCollidable(boolean ignoreClimbing) {
+        if (!this.isLobotomized() || !this.passengers.isEmpty() || this.lastHurtByPlayer != null || this.lastHurtByMob != null || net.minecraft.server.MinecraftServer.currentTick % 100 < 4) {
+            return super.isCollidable(ignoreClimbing);
+        }
+        return false;
+    }
+    // KioCG end
+
     @Override
     public Brain<Villager> getBrain() {
         return (Brain<Villager>) super.getBrain(); // CraftBukkit - decompile error
@@ -346,14 +364,14 @@ public class Villager extends AbstractVillager implements ReputationEventHandler
         // Purpur start
         if (this.level().purpurConfig.villagerLobotomizeEnabled) {
             // treat as inactive if lobotomized
-            inactive = inactive || checkLobotomized();
+            inactive = this.lobotomized.checkLobotomized(this) || inactive; // KioCG - 使用更好的检测
         } else {
             // clean up state for API
             this.isLobotomized = false;
         }
         if (!inactive && (getRider() == null || !this.isControllable()) && this.behaviorTick++ % this.activatedPriority == 0) // Purpur - only use brain if no rider
             this.getBrain().tick((ServerLevel) this.level(), this); // Paper
-        else if (this.isLobotomized && shouldRestock()) restock();
+        else if (this.lobotomized.isLobotomized() && shouldRestock()) restock(); // KioCG - 使用更好的检测
         // Purpur end
         //this.level().getProfiler().pop(); // Purpur
         if (this.assignProfessionWhenSpawned) {
@@ -658,6 +676,8 @@ public class Villager extends AbstractVillager implements ReputationEventHandler
             nbt.putBoolean("AssignProfessionWhenSpawned", true);
         }
 
+        nbt.putBoolean("KioCG.Lobotomized", this.lobotomized.isLobotomized()); // KioCG
+
     }
 
     @Override
@@ -698,6 +718,14 @@ public class Villager extends AbstractVillager implements ReputationEventHandler
             this.assignProfessionWhenSpawned = nbt.getBoolean("AssignProfessionWhenSpawned");
         }
 
+        // KioCG start
+        boolean isLobotomized = nbt.getBoolean("KioCG.Lobotomized");
+        this.lobotomized.setLobotomized(isLobotomized);
+        if (isLobotomized) {
+            this.onGround = true;
+        }
+        // KioCG end
+
     }
 
     @Override
