From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Joket_abc <15005771700@qq.com>
Date: Wed, 16 Nov 2022 06:16:03 +0800
Subject: [PATCH] Linear achievement system


diff --git a/src/main/java/net/minecraft/server/PlayerAdvancements.java b/src/main/java/net/minecraft/server/PlayerAdvancements.java
index 3919d9c193abcfd8b97dfb0ceb38638440f60fbe..198247790b746e4db115f3131f977d6e0e38767c 100644
--- a/src/main/java/net/minecraft/server/PlayerAdvancements.java
+++ b/src/main/java/net/minecraft/server/PlayerAdvancements.java
@@ -237,6 +237,13 @@ public class PlayerAdvancements {
             }
             // Paper end
             this.unregisterListeners(advancement);
+            // KioCG start - 线性的成就系统
+            Iterator iterator = advancement.getChildren().iterator();
+            while (iterator.hasNext()) {
+                Advancement children = (Advancement) iterator.next();
+                this.registerListenersLinear(children);
+            }
+            // KioCG end
             this.progressChanged.add(advancement);
             flag = true;
             if (!flag1 && advancementprogress.isDone()) {
@@ -287,6 +294,18 @@ public class PlayerAdvancements {
     }
 
     private void registerListeners(Advancement advancement) {
+    // KioCG start - 线性的成就系统
+        if (advancement.getDisplay() == null || advancement.getParent() == null || this.getOrStartProgress(advancement.getParent()).isDone()) {
+            this.registerListenersLinear(advancement);
+            return;
+        }
+        String path = advancement.getId().getPath();
+        if (path.startsWith("bacap") || path.startsWith("challenges")) {
+            this.registerListenersLinear(advancement);
+        }
+    }
+    private void registerListenersLinear(Advancement advancement) {
+    // KioCG end - 线性的成就系统
         AdvancementProgress advancementprogress = this.getOrStartProgress(advancement);
 
         if (!advancementprogress.isDone()) {
