From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Joket_abc <15005771700@qq.com>
Date: Fri, 13 Oct 2023 21:44:33 +0800
Subject: [PATCH] Linear achievement system


diff --git a/src/main/java/net/minecraft/server/PlayerAdvancements.java b/src/main/java/net/minecraft/server/PlayerAdvancements.java
index 533a348acb53c8e7590f3d81c19e53a655101713..4f55ebdad8f23fcb3b2784a7b3fc40b950d0d3b5 100644
--- a/src/main/java/net/minecraft/server/PlayerAdvancements.java
+++ b/src/main/java/net/minecraft/server/PlayerAdvancements.java
@@ -235,6 +235,14 @@ public class PlayerAdvancements {
             }
             // Paper end
             this.unregisterListeners(advancement);
+            // KioCG start - 线性的成就系统
+            final AdvancementNode advancementNode = MinecraftServer.getServer().getAdvancements().tree().get(advancement);
+            if (advancementNode != null) {
+                for (final AdvancementNode child : advancementNode.children()) {
+                    this.registerListenersLinear(child.holder());
+                }
+            }
+            // KioCG end
             this.progressChanged.add(advancement);
             flag = true;
             if (!flag1 && advancementprogress.isDone()) {
@@ -291,6 +299,27 @@ public class PlayerAdvancements {
     }
 
     private void registerListeners(AdvancementHolder advancement) {
+    // KioCG start - 线性的成就系统
+        if (advancement.value().display().isEmpty()) {
+            this.registerListenersLinear(advancement);
+        } else if (advancement.value().isRoot()) {
+            this.registerListenersLinear(advancement);
+        } else {
+            AdvancementHolder parent = advancement.value().parent().map(MinecraftServer.getServer().getAdvancements()::get).orElseThrow();
+            if (this.getOrStartProgress(parent).isDone()) {
+                this.registerListenersLinear(advancement);
+            } else if (parent.value().display().isPresent() && parent.value().display().get().isHidden()) {
+                this.registerListenersLinear(advancement);
+            } else {
+                String path = advancement.id().getPath();
+                if (path.startsWith("bacap") || path.startsWith("challenges")) {
+                    this.registerListenersLinear(advancement);
+                }
+            }
+        }
+    }
+    private void registerListenersLinear(AdvancementHolder advancement) {
+    // KioCG end - 线性的成就系统
         AdvancementProgress advancementprogress = this.getOrStartProgress(advancement);
 
         if (!advancementprogress.isDone()) {
