From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Joket_abc <15005771700@qq.com>
Date: Wed, 16 Nov 2022 22:37:53 +0800
Subject: [PATCH] Shears can have silk touch enchantment


diff --git a/src/main/java/net/minecraft/world/item/enchantment/EnchantmentCategory.java b/src/main/java/net/minecraft/world/item/enchantment/EnchantmentCategory.java
index fc2c35f57436371cb0111aedfd289ac95d506d07..a09bd528218d44630300fcf8f263e7b164d7051d 100644
--- a/src/main/java/net/minecraft/world/item/enchantment/EnchantmentCategory.java
+++ b/src/main/java/net/minecraft/world/item/enchantment/EnchantmentCategory.java
@@ -135,6 +135,14 @@ public enum EnchantmentCategory {
             return WEAPON.canEnchant(item) || item instanceof net.minecraft.world.item.ShearsItem;
         }
     // Purpur end
+    // KioCG start
+    },
+    TOOL_AND_SHEARS {
+        @Override
+        public boolean canEnchant(Item item) {
+            return DIGGER.canEnchant(item) || item instanceof net.minecraft.world.item.ShearsItem;
+        }
+    // KioCG end
     };
 
     public abstract boolean canEnchant(Item item);
diff --git a/src/main/java/net/minecraft/world/item/enchantment/UntouchingEnchantment.java b/src/main/java/net/minecraft/world/item/enchantment/UntouchingEnchantment.java
index c4245ecf24066c708beb434a3b610fc7c6b07044..84550504d647224cb9e8559cf40aa6e98af10dc8 100644
--- a/src/main/java/net/minecraft/world/item/enchantment/UntouchingEnchantment.java
+++ b/src/main/java/net/minecraft/world/item/enchantment/UntouchingEnchantment.java
@@ -7,6 +7,13 @@ public class UntouchingEnchantment extends Enchantment {
         super(weight, EnchantmentCategory.DIGGER, slotTypes);
     }
 
+    // KioCG start
+    @Override
+    public boolean canEnchant(net.minecraft.world.item.ItemStack stack) {
+        return EnchantmentCategory.TOOL_AND_SHEARS.canEnchant(stack.getItem());
+    }
+    // KioCG end
+
     @Override
     public int getMinCost(int level) {
         return 15;
