From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Joket_abc <15005771700@qq.com>
Date: Thu, 15 Dec 2022 00:04:23 +0800
Subject: [PATCH] Prevent fire spreading


diff --git a/src/main/java/net/minecraft/world/level/block/FireBlock.java b/src/main/java/net/minecraft/world/level/block/FireBlock.java
index 7f05e58358024d303eab9ab4fbc1bb299760ad1e..785a1045a447539e17608dde56ddd7c4652e6d27 100644
--- a/src/main/java/net/minecraft/world/level/block/FireBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/FireBlock.java
@@ -219,6 +219,8 @@ public class FireBlock extends BaseFireBlock {
                 this.trySpread(world, pos.north(), 300 + k, random, i, pos);
                 this.trySpread(world, pos.south(), 300 + k, random, i, pos);
                 // CraftBukkit end
+
+                if (true) return; // KioCG - 阻止火焰蔓延
                 BlockPos.MutableBlockPos blockposition_mutableblockposition = new BlockPos.MutableBlockPos();
 
                 for (int l = -1; l <= 1; ++l) {
diff --git a/src/main/java/net/minecraft/world/level/material/LavaFluid.java b/src/main/java/net/minecraft/world/level/material/LavaFluid.java
index 2d492d849ff73a738dfbcb16507feb89bf19a962..cf74843b241c5adb3f6c5623cc90cdd2dd08f932 100644
--- a/src/main/java/net/minecraft/world/level/material/LavaFluid.java
+++ b/src/main/java/net/minecraft/world/level/material/LavaFluid.java
@@ -68,6 +68,7 @@ public abstract class LavaFluid extends FlowingFluid {
 
     @Override
     public void randomTick(Level world, BlockPos pos, FluidState state, RandomSource random) {
+        if (net.minecraft.server.MinecraftServer.currentTick % 10 == 0) // KioCG
         if (world.getGameRules().getBoolean(GameRules.RULE_DOFIRETICK)) {
             int i = random.nextInt(3);
 
