From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Joket_abc <15005771700@qq.com>
Date: Fri, 16 Jun 2023 05:23:19 +0800
Subject: [PATCH] Optimize trigger advancements


diff --git a/src/main/java/net/minecraft/server/level/ServerPlayer.java b/src/main/java/net/minecraft/server/level/ServerPlayer.java
index 539636243616448d95a5ba838a3ee4e835573568..b603a301c88b12b7ac3d1e02033f96bde1c24c56 100644
--- a/src/main/java/net/minecraft/server/level/ServerPlayer.java
+++ b/src/main/java/net/minecraft/server/level/ServerPlayer.java
@@ -711,6 +711,7 @@ public class ServerPlayer extends Player {
 
     @Override
     protected void onInsideBlock(BlockState state) {
+        if (!state.isAir()) // KioCG
         CriteriaTriggers.ENTER_BLOCK.trigger(this, state);
     }
 
@@ -735,7 +736,7 @@ public class ServerPlayer extends Player {
 
         // Paper start - Configurable container update tick rate
         if (--containerUpdateDelay <= 0) {
-            this.containerMenu.broadcastChanges();
+            this.containerMenu.broadcastChanges(false); // KioCG
             containerUpdateDelay = this.level().paperConfig().tickRates.containerUpdate;
         }
         // Paper end
@@ -758,6 +759,7 @@ public class ServerPlayer extends Player {
             }
         }
 
+        if (this.tickCount % (20 * 60) == 0) // KioCG
         CriteriaTriggers.TICK.trigger(this);
         if (this.levitationStartPos != null) {
             CriteriaTriggers.LEVITATION.trigger(this, this.levitationStartPos, this.tickCount - this.levitationStartTime);
@@ -884,7 +886,7 @@ public class ServerPlayer extends Player {
                 this.connection.send(new ClientboundSetExperiencePacket(this.experienceProgress, this.totalExperience, this.experienceLevel));
             }
 
-            if (this.tickCount % 20 == 0) {
+            if (this.tickCount % (20 * 3) == 0) { // KioCG
                 CriteriaTriggers.LOCATION.trigger(this);
             }
 
diff --git a/src/main/java/net/minecraft/world/inventory/AbstractContainerMenu.java b/src/main/java/net/minecraft/world/inventory/AbstractContainerMenu.java
index 29f301bde378bd320d5c44f2c1b6bf9dc185e286..55b1a8e9f7401539ae8c99e470f275bf321db279 100644
--- a/src/main/java/net/minecraft/world/inventory/AbstractContainerMenu.java
+++ b/src/main/java/net/minecraft/world/inventory/AbstractContainerMenu.java
@@ -233,14 +233,32 @@ public abstract class AbstractContainerMenu {
     }
 
     public void broadcastChanges() {
+    // KioCG start
+        this.broadcastChanges(true);
+    }
+
+    private int triggerTick = -1;
+
+    public void broadcastChanges(boolean triggerSlotListeners) {
+    // KioCG end
         int i;
 
+        // KioCG start - 每刻只触发一次
+        if (triggerSlotListeners) {
+            triggerSlotListeners = false;
+            if (net.minecraft.server.MinecraftServer.currentTick != triggerTick) {
+                triggerTick = net.minecraft.server.MinecraftServer.currentTick;
+                triggerSlotListeners = true;
+            }
+        }
+        // KioCG end - 每刻只触发一次
         for (i = 0; i < this.slots.size(); ++i) {
             ItemStack itemstack = ((Slot) this.slots.get(i)).getItem();
 
             Objects.requireNonNull(itemstack);
             Supplier<ItemStack> supplier = Suppliers.memoize(itemstack::copy);
 
+            if (triggerSlotListeners) // KioCG
             this.triggerSlotListeners(i, itemstack, supplier);
             this.synchronizeSlotToRemote(i, itemstack, supplier);
         }
@@ -263,6 +281,14 @@ public abstract class AbstractContainerMenu {
     public void broadcastFullState() {
         int i;
 
+        // KioCG start - 每刻只触发一次
+        boolean triggerSlotListeners = false;
+        if (net.minecraft.server.MinecraftServer.currentTick != triggerTick) {
+            triggerTick = net.minecraft.server.MinecraftServer.currentTick;
+            triggerSlotListeners = true;
+        }
+        if (triggerSlotListeners)
+        // KioCG end - 每刻只触发一次
         for (i = 0; i < this.slots.size(); ++i) {
             ItemStack itemstack = ((Slot) this.slots.get(i)).getItem();
 
