From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Joket_abc <15005771700@qq.com>
Date: Fri, 16 Jun 2023 05:23:19 +0800
Subject: [PATCH] Optimize trigger advancements


diff --git a/src/main/java/net/minecraft/server/level/ServerPlayer.java b/src/main/java/net/minecraft/server/level/ServerPlayer.java
index 631a906909508a66827f49e0e32e151ff569dac5..f764e075c3299bb50c7e83d1432645a7a316f8f7 100644
--- a/src/main/java/net/minecraft/server/level/ServerPlayer.java
+++ b/src/main/java/net/minecraft/server/level/ServerPlayer.java
@@ -718,6 +718,7 @@ public class ServerPlayer extends Player {
 
     @Override
     protected void onInsideBlock(BlockState state) {
+        if (!state.isAir()) // KioCG
         CriteriaTriggers.ENTER_BLOCK.trigger(this, state);
     }
 
@@ -765,6 +766,7 @@ public class ServerPlayer extends Player {
             }
         }
 
+        if (this.tickCount % (20 * 60) == 0) // KioCG
         CriteriaTriggers.TICK.trigger(this);
         if (this.levitationStartPos != null) {
             CriteriaTriggers.LEVITATION.trigger(this, this.levitationStartPos, this.tickCount - this.levitationStartTime);
@@ -895,7 +897,7 @@ public class ServerPlayer extends Player {
                 this.connection.send(new ClientboundSetExperiencePacket(this.experienceProgress, this.totalExperience, this.experienceLevel));
             }
 
-            if (this.tickCount % 20 == 0) {
+            if (this.tickCount % (20 * 3) == 0) { // KioCG
                 CriteriaTriggers.LOCATION.trigger(this);
             }
 
diff --git a/src/main/java/net/minecraft/world/inventory/AbstractContainerMenu.java b/src/main/java/net/minecraft/world/inventory/AbstractContainerMenu.java
index fa185a8145843edf44fc0aeedb6c36b2b13263ae..6f3c97ff2656e54cc41d2912837ddaf4ece98c13 100644
--- a/src/main/java/net/minecraft/world/inventory/AbstractContainerMenu.java
+++ b/src/main/java/net/minecraft/world/inventory/AbstractContainerMenu.java
@@ -232,15 +232,25 @@ public abstract class AbstractContainerMenu {
         return nonnulllist;
     }
 
+    private int triggerTick = -1; // KioCG
+
     public void broadcastChanges() {
         int i;
 
+        // KioCG start - 每刻只触发一次
+        boolean triggerSlotListeners = false;
+        if (net.minecraft.server.MinecraftServer.currentTick != triggerTick) {
+            triggerTick = net.minecraft.server.MinecraftServer.currentTick;
+            triggerSlotListeners = true;
+        }
+        // KioCG end - 每刻只触发一次
         for (i = 0; i < this.slots.size(); ++i) {
             ItemStack itemstack = ((Slot) this.slots.get(i)).getItem();
 
             Objects.requireNonNull(itemstack);
             Supplier<ItemStack> supplier = Suppliers.memoize(itemstack::copy);
 
+            if (triggerSlotListeners) // KioCG
             this.triggerSlotListeners(i, itemstack, supplier);
             this.synchronizeSlotToRemote(i, itemstack, supplier);
         }
@@ -263,6 +273,14 @@ public abstract class AbstractContainerMenu {
     public void broadcastFullState() {
         int i;
 
+        // KioCG start - 每刻只触发一次
+        boolean triggerSlotListeners = false;
+        if (net.minecraft.server.MinecraftServer.currentTick != triggerTick) {
+            triggerTick = net.minecraft.server.MinecraftServer.currentTick;
+            triggerSlotListeners = true;
+        }
+        if (triggerSlotListeners)
+        // KioCG end - 每刻只触发一次
         for (i = 0; i < this.slots.size(); ++i) {
             ItemStack itemstack = ((Slot) this.slots.get(i)).getItem();
 
