From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Joket_abc <15005771700@qq.com>
Date: Fri, 16 Jun 2023 05:23:19 +0800
Subject: [PATCH] Optimize trigger advancements


diff --git a/src/main/java/net/minecraft/server/level/ServerPlayer.java b/src/main/java/net/minecraft/server/level/ServerPlayer.java
index 5752881da18886311d3a8b0721e0dda2e85f5021..e9c2440b9551ec69ed2d4d9917d1b7def031fcaa 100644
--- a/src/main/java/net/minecraft/server/level/ServerPlayer.java
+++ b/src/main/java/net/minecraft/server/level/ServerPlayer.java
@@ -708,6 +708,7 @@ public class ServerPlayer extends Player {
 
     @Override
     protected void onInsideBlock(BlockState state) {
+        if (!state.isAir()) // KioCG
         CriteriaTriggers.ENTER_BLOCK.trigger(this, state);
     }
 
@@ -732,7 +733,7 @@ public class ServerPlayer extends Player {
 
         // Paper start - Configurable container update tick rate
         if (--containerUpdateDelay <= 0) {
-            this.containerMenu.broadcastChanges();
+            this.containerMenu.broadcastChanges(false); // KioCG
             containerUpdateDelay = this.level().paperConfig().tickRates.containerUpdate;
         }
         // Paper end
@@ -755,6 +756,7 @@ public class ServerPlayer extends Player {
             }
         }
 
+        if (this.tickCount % (20 * 60) == 0) // KioCG
         CriteriaTriggers.TICK.trigger(this);
         if (this.levitationStartPos != null) {
             CriteriaTriggers.LEVITATION.trigger(this, this.levitationStartPos, this.tickCount - this.levitationStartTime);
@@ -871,7 +873,7 @@ public class ServerPlayer extends Player {
                 this.connection.send(new ClientboundSetExperiencePacket(this.experienceProgress, this.totalExperience, this.experienceLevel));
             }
 
-            if (this.tickCount % 20 == 0) {
+            if (this.tickCount % (20 * 3) == 0) { // KioCG
                 CriteriaTriggers.LOCATION.trigger(this);
             }
 
diff --git a/src/main/java/net/minecraft/world/inventory/AbstractContainerMenu.java b/src/main/java/net/minecraft/world/inventory/AbstractContainerMenu.java
index 9ca261c9f21279558961649cb4849ac379d67573..9be1da572516b59b83ae613dcbf8c1858c7f7b1c 100644
--- a/src/main/java/net/minecraft/world/inventory/AbstractContainerMenu.java
+++ b/src/main/java/net/minecraft/world/inventory/AbstractContainerMenu.java
@@ -233,6 +233,12 @@ public abstract class AbstractContainerMenu {
     }
 
     public void broadcastChanges() {
+    // KioCG start
+        this.broadcastChanges(true);
+    }
+
+    public void broadcastChanges(boolean triggerSlotListeners) {
+    // KioCG end
         int i;
 
         for (i = 0; i < this.slots.size(); ++i) {
@@ -241,6 +247,7 @@ public abstract class AbstractContainerMenu {
             Objects.requireNonNull(itemstack);
             Supplier<ItemStack> supplier = Suppliers.memoize(itemstack::copy);
 
+            if (triggerSlotListeners) // KioCG
             this.triggerSlotListeners(i, itemstack, supplier);
             this.synchronizeSlotToRemote(i, itemstack, supplier);
         }
@@ -292,7 +299,15 @@ public abstract class AbstractContainerMenu {
 
     }
 
+    private int triggerTick = -1; // KioCG
     private void triggerSlotListeners(int slot, ItemStack stack, Supplier<ItemStack> copySupplier) {
+        // KioCG start - 每刻只触发一次
+        if (net.minecraft.server.MinecraftServer.currentTick == triggerTick) {
+            return;
+        }
+        triggerTick = net.minecraft.server.MinecraftServer.currentTick;
+        // KioCG end - 每刻只触发一次
+
         ItemStack itemstack1 = (ItemStack) this.lastSlots.get(slot);
 
         if (!ItemStack.matches(itemstack1, stack)) {
