From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Joket_abc <15005771700@qq.com>
Date: Mon, 2 Jan 2023 04:27:42 +0800
Subject: [PATCH] Reduce item flow speed


diff --git a/src/main/java/net/minecraft/world/entity/item/ItemEntity.java b/src/main/java/net/minecraft/world/entity/item/ItemEntity.java
index 9c3db8f774e5c11df18d2c317c874e8ac26e7f8e..3f39661eb002a2d8621d85cf715231e87b3e9756 100644
--- a/src/main/java/net/minecraft/world/entity/item/ItemEntity.java
+++ b/src/main/java/net/minecraft/world/entity/item/ItemEntity.java
@@ -236,7 +236,7 @@ public class ItemEntity extends Entity {
     private void setUnderwaterMovement() {
         Vec3 vec3d = this.getDeltaMovement();
 
-        this.setDeltaMovement(vec3d.x * 0.9900000095367432D, vec3d.y + (double) (vec3d.y < 0.05999999865889549D ? 5.0E-4F : 0.0F), vec3d.z * 0.9900000095367432D);
+        this.setDeltaMovement(vec3d.x * 0.7900000214576721D, vec3d.y + (double) (vec3d.y < 0.05999999865889549D ? 5.0E-4F : 0.0F), vec3d.z * 0.7900000214576721D); // KioCG - 降低物品流速
     }
 
     private void setUnderLavaMovement() {
