From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Joket_abc <15005771700@qq.com>
Date: Sun, 2 Apr 2023 07:49:16 +0800
Subject: [PATCH] Remove Not Secure


diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
index 7cd537847c701fd2de1f8c7bbacf38e71c499720..570c9072be8ce4d7eb0338b06ce465d5194cbd2e 100644
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -2661,7 +2661,7 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
         // Paper start
         net.kyori.adventure.text.Component s1 = io.papermc.paper.adventure.PaperAdventure.asAdventure(params.decorate(message));
 
-        if (prefix != null) {
+        if (prefix != null && !"Not Secure".equals(prefix)) { // KioCG
             MinecraftServer.COMPONENT_LOGGER.info("[{}] {}", prefix, s1);
         } else {
             MinecraftServer.COMPONENT_LOGGER.info("{}", s1);
