From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Joket_abc <15005771700@qq.com>
Date: Thu, 23 Feb 2023 05:20:59 +0800
Subject: [PATCH] Remove Not Secure


diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
index ace34fa05c0e583cea352a0ad9995d6ebb1f97a1..ee7a6a305633b1e9b2b7beff94826fdd1f3f1516 100644
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -2663,7 +2663,7 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
     public void logChatMessage(Component message, ChatType.Bound params, @Nullable String prefix) {
         String s1 = params.decorate(message).getString();
 
-        if (prefix != null) {
+        if (prefix != null && !"Not Secure".equals(prefix)) { // KioCG
             MinecraftServer.LOGGER.info("[{}] {}", prefix, s1);
         } else {
             MinecraftServer.LOGGER.info("{}", s1);
