From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Joket_abc <15005771700@qq.com>
Date: Thu, 23 Feb 2023 05:20:59 +0800
Subject: [PATCH] Remove Not Secure


diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
index 393cede8c79da3082f8f0028f728d3b8f6f03167..12d04343940e8a81ed493c2d79215522989f2b78 100644
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -2638,7 +2638,7 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
     public void logChatMessage(Component message, ChatType.Bound params, @Nullable String prefix) {
         String s1 = params.decorate(message).getString();
 
-        if (prefix != null) {
+        if (prefix != null && !"Not Secure".equals(prefix)) { // KioCG
             MinecraftServer.LOGGER.info("[{}] {}", prefix, s1);
         } else {
             MinecraftServer.LOGGER.info("{}", s1);
