From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Joket_abc <15005771700@qq.com>
Date: Thu, 23 Feb 2023 05:20:59 +0800
Subject: [PATCH] Remove Not Secure


diff --git a/src/main/java/net/minecraft/server/MinecraftServer.java b/src/main/java/net/minecraft/server/MinecraftServer.java
index 310b3acb490f839a5f433371781d004e6a13686e..276952a880f246040bed58e614107bdc18413853 100644
--- a/src/main/java/net/minecraft/server/MinecraftServer.java
+++ b/src/main/java/net/minecraft/server/MinecraftServer.java
@@ -2659,7 +2659,7 @@ public abstract class MinecraftServer extends ReentrantBlockableEventLoop<TickTa
     public void logChatMessage(Component message, ChatType.Bound params, @Nullable String prefix) {
         String s1 = params.decorate(message).getString();
 
-        if (prefix != null) {
+        if (prefix != null && !"Not Secure".equals(prefix)) { // KioCG
             MinecraftServer.LOGGER.info("[{}] {}", prefix, s1);
         } else {
             MinecraftServer.LOGGER.info("{}", s1);
