From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Joket_abc <15005771700@qq.com>
Date: Thu, 12 Oct 2023 21:00:29 +0800
Subject: [PATCH] Configurable can drop mobs skull


diff --git a/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java b/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java
index ee6c4ac72efd674c3a35fcbec5b9f363f2f5413d..f7f1c3aab99bc7e6aa3f46855675f819cfd6c809 100644
--- a/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java
+++ b/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java
@@ -556,6 +556,20 @@ public class WorldConfiguration extends ConfigurationPart {
             public long phaseOne = 300000L;
             public long phaseTwo = 700000L;
         }
+
+        public Entities entities;
+
+        public class Entities extends ConfigurationPart {
+            public Creeper creeper;
+
+            public class Creeper extends ConfigurationPart {
+                public boolean canDropCreeperSkull = true;
+                public boolean canDropSkeletonSkull = true;
+                public boolean canDropWitherSkeletonSkull = true;
+                public boolean canDropZombieSkull = true;
+                public boolean canDropPiglinSkull = true;
+            }
+        }
     }
     // KioCG end
 }
diff --git a/src/main/java/net/minecraft/world/entity/monster/Creeper.java b/src/main/java/net/minecraft/world/entity/monster/Creeper.java
index 8bbc11845d190b33296f61c503f54209d218bc85..ea270b42204ffedc8e486ac1f0462251d890c19b 100644
--- a/src/main/java/net/minecraft/world/entity/monster/Creeper.java
+++ b/src/main/java/net/minecraft/world/entity/monster/Creeper.java
@@ -300,7 +300,7 @@ public class Creeper extends Monster implements PowerableMob {
         if (entity != this && entity instanceof Creeper) {
             Creeper entitycreeper = (Creeper) entity;
 
-            if (entitycreeper.canDropMobsSkull()) {
+            if (entitycreeper.level().paperConfig().kiocgConfig.entities.creeper.canDropCreeperSkull && entitycreeper.canDropMobsSkull()) { // KioCG
                 entitycreeper.increaseDroppedSkulls();
                 this.spawnAtLocation((ItemLike) Items.CREEPER_HEAD);
             }
diff --git a/src/main/java/net/minecraft/world/entity/monster/Skeleton.java b/src/main/java/net/minecraft/world/entity/monster/Skeleton.java
index 6df5d100f21f26d88b3a36768bc9220f64d47cb8..ede47e5da73097fec9435f9839646bf9df93336b 100644
--- a/src/main/java/net/minecraft/world/entity/monster/Skeleton.java
+++ b/src/main/java/net/minecraft/world/entity/monster/Skeleton.java
@@ -177,7 +177,7 @@ public class Skeleton extends AbstractSkeleton {
         if (entity instanceof Creeper) {
             Creeper entitycreeper = (Creeper) entity;
 
-            if (entitycreeper.canDropMobsSkull()) {
+            if (entitycreeper.level().paperConfig().kiocgConfig.entities.creeper.canDropSkeletonSkull && entitycreeper.canDropMobsSkull()) { // KioCG
                 entitycreeper.increaseDroppedSkulls();
                 this.spawnAtLocation((ItemLike) Items.SKELETON_SKULL);
             }
diff --git a/src/main/java/net/minecraft/world/entity/monster/WitherSkeleton.java b/src/main/java/net/minecraft/world/entity/monster/WitherSkeleton.java
index e97cb4e166c2e9ac6d93ed5b15350758326e7e74..fd8b629431037af8c1dffed171eaa050177cf231 100644
--- a/src/main/java/net/minecraft/world/entity/monster/WitherSkeleton.java
+++ b/src/main/java/net/minecraft/world/entity/monster/WitherSkeleton.java
@@ -101,7 +101,7 @@ public class WitherSkeleton extends AbstractSkeleton {
         if (entity instanceof Creeper) {
             Creeper entitycreeper = (Creeper) entity;
 
-            if (entitycreeper.canDropMobsSkull()) {
+            if (entitycreeper.level().paperConfig().kiocgConfig.entities.creeper.canDropWitherSkeletonSkull && entitycreeper.canDropMobsSkull()) { // KioCG
                 entitycreeper.increaseDroppedSkulls();
                 this.spawnAtLocation((ItemLike) Items.WITHER_SKELETON_SKULL);
             }
diff --git a/src/main/java/net/minecraft/world/entity/monster/Zombie.java b/src/main/java/net/minecraft/world/entity/monster/Zombie.java
index 4e69136a9316a2f8a67d4da9e795076f0d5943f7..a9cf91a406f13cabc93e3292a643db04f900667d 100644
--- a/src/main/java/net/minecraft/world/entity/monster/Zombie.java
+++ b/src/main/java/net/minecraft/world/entity/monster/Zombie.java
@@ -644,7 +644,7 @@ public class Zombie extends Monster {
         if (entity instanceof Creeper) {
             Creeper entitycreeper = (Creeper) entity;
 
-            if (entitycreeper.canDropMobsSkull()) {
+            if (entitycreeper.level().paperConfig().kiocgConfig.entities.creeper.canDropZombieSkull && entitycreeper.canDropMobsSkull()) { // KioCG
                 ItemStack itemstack = this.getSkull();
 
                 if (!itemstack.isEmpty()) {
diff --git a/src/main/java/net/minecraft/world/entity/monster/piglin/Piglin.java b/src/main/java/net/minecraft/world/entity/monster/piglin/Piglin.java
index 1422c0f4ff6a3e61f229574cd7b50971bdbd8451..9b472e2bbeec8fd03c5371bd6283942b4946d138 100644
--- a/src/main/java/net/minecraft/world/entity/monster/piglin/Piglin.java
+++ b/src/main/java/net/minecraft/world/entity/monster/piglin/Piglin.java
@@ -176,7 +176,7 @@ public class Piglin extends AbstractPiglin implements CrossbowAttackMob, Invento
         if (entity instanceof Creeper) {
             Creeper entitycreeper = (Creeper) entity;
 
-            if (entitycreeper.canDropMobsSkull()) {
+            if (entitycreeper.level().paperConfig().kiocgConfig.entities.creeper.canDropPiglinSkull && entitycreeper.canDropMobsSkull()) { // KioCG
                 ItemStack itemstack = new ItemStack(Items.PIGLIN_HEAD);
 
                 entitycreeper.increaseDroppedSkulls();
