From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Joket_abc <15005771700@qq.com>
Date: Mon, 3 Apr 2023 10:25:37 +0800
Subject: [PATCH] Balanced EPF value


diff --git a/src/main/java/net/minecraft/world/damagesource/CombatRules.java b/src/main/java/net/minecraft/world/damagesource/CombatRules.java
index aa2331c6df4e79d4bb0add071a0b11d2a3a08b88..f5cb00a39c5b7f621fa53c665427c307f6cb6fbe 100644
--- a/src/main/java/net/minecraft/world/damagesource/CombatRules.java
+++ b/src/main/java/net/minecraft/world/damagesource/CombatRules.java
@@ -16,7 +16,7 @@ public class CombatRules {
     }
 
     public static float getDamageAfterMagicAbsorb(float damageDealt, float protection) {
-        float f = Mth.clamp(protection, 0.0F, org.purpurmc.purpur.PurpurConfig.limitArmor ? 20F : Float.MAX_VALUE); // Purpur
-        return damageDealt * (1.0F - f / 25.0F);
+        float f = Mth.clamp(protection, 0.0F, org.purpurmc.purpur.PurpurConfig.limitArmor ? 40F : Float.MAX_VALUE); // Purpur // KioCG - Balanced EPF value
+        return damageDealt * (1.0F - f / 50.0F); // KioCG - Balanced EPF value
     }
 }
