From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Joket_abc <15005771700@qq.com>
Date: Mon, 16 Oct 2023 03:40:49 +0800
Subject: [PATCH] Configurable persistent when picked up items


diff --git a/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java b/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java
index f7f1c3aab99bc7e6aa3f46855675f819cfd6c809..e7f35737769dc907dbf227fb7c696a5386644c68 100644
--- a/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java
+++ b/src/main/java/io/papermc/paper/configuration/WorldConfiguration.java
@@ -569,6 +569,8 @@ public class WorldConfiguration extends ConfigurationPart {
                 public boolean canDropZombieSkull = true;
                 public boolean canDropPiglinSkull = true;
             }
+
+            public boolean persistentOnlyWhenPickedUpEnchantedItems = false;
         }
     }
     // KioCG end
diff --git a/src/main/java/net/minecraft/world/entity/Mob.java b/src/main/java/net/minecraft/world/entity/Mob.java
index a2588ffd854a4610cd24942b1fea0dc5080a511b..5dd03e2535de45cb06d15ff2cb4f174106e6bbc6 100644
--- a/src/main/java/net/minecraft/world/entity/Mob.java
+++ b/src/main/java/net/minecraft/world/entity/Mob.java
@@ -392,6 +392,7 @@ public abstract class Mob extends LivingEntity implements Targeting {
         }
         if (this.ticksSinceLastInteraction > this.level().purpurConfig.entityLifeSpan) {
             this.discard();
+            if (this.level().paperConfig().kiocgConfig.entities.persistentOnlyWhenPickedUpEnchantedItems) this.dropPickedUpItems(); // KioCG
         }
     }
     // Purpur end
@@ -795,9 +796,32 @@ public abstract class Mob extends LivingEntity implements Targeting {
     protected void setItemSlotAndDropWhenKilled(EquipmentSlot slot, ItemStack stack) {
         this.setItemSlot(slot, stack);
         this.setGuaranteedDrop(slot);
+        if (!this.level().paperConfig().kiocgConfig.entities.persistentOnlyWhenPickedUpEnchantedItems || stack.isEnchanted()) // KioCG
         this.persistenceRequired = true;
     }
 
+    // KioCG start
+    private void dropPickedUpItems() {
+        if (this.canPickUpLoot()) {
+            EquipmentSlot[] aenumitemslot = EquipmentSlot.values();
+            int j = aenumitemslot.length;
+            for (int k = 0; k < j; ++k) {
+                EquipmentSlot enumitemslot = aenumitemslot[k];
+                if (this.getEquipmentDropChance(enumitemslot) != 2.0f) {
+                    continue;
+                }
+
+                ItemStack itemstack = this.getItemBySlot(enumitemslot);
+                if (!itemstack.isEmpty() && !EnchantmentHelper.hasVanishingCurse(itemstack)) {
+                    this.forceDrops = true;
+                    this.spawnAtLocation(itemstack);
+                    this.forceDrops = false;
+                }
+            }
+        }
+    }
+    // KioCG end
+
     public void setGuaranteedDrop(EquipmentSlot slot) {
         switch (slot.getType()) {
             case HAND:
@@ -907,6 +931,7 @@ public abstract class Mob extends LivingEntity implements Targeting {
 
                 if (d0 > (double) j && this.removeWhenFarAway(d0)) {
                     this.discard();
+                    if (this.level().paperConfig().kiocgConfig.entities.persistentOnlyWhenPickedUpEnchantedItems) this.dropPickedUpItems(); // KioCG
                 }
 
                 int k = this.level().paperConfig().entities.spawning.despawnRanges.get(this.getType().getCategory()).soft(); // Paper - Configurable despawn distances
@@ -914,6 +939,7 @@ public abstract class Mob extends LivingEntity implements Targeting {
 
                 if (this.noActionTime > 600 && this.random.nextInt(800) == 0 && d0 > (double) l && this.removeWhenFarAway(d0)) {
                     this.discard();
+                    if (this.level().paperConfig().kiocgConfig.entities.persistentOnlyWhenPickedUpEnchantedItems) this.dropPickedUpItems(); // KioCG
                 } else if (d0 < (double) l) {
                     this.noActionTime = 0;
                 }
