From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Joket_abc <15005771700@qq.com>
Date: Wed, 3 May 2023 19:17:00 +0800
Subject: [PATCH] New agricultural system


diff --git a/src/main/java/net/minecraft/world/level/block/CropBlock.java b/src/main/java/net/minecraft/world/level/block/CropBlock.java
index f1016932cba6905c9cd474daeae648b2dcc1f32d..4479c981ef6e3d9187dd1c30027d9f02b4d6305c 100644
--- a/src/main/java/net/minecraft/world/level/block/CropBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/CropBlock.java
@@ -94,7 +94,17 @@ public class CropBlock extends BushBlock implements BonemealableBlock {
 
                 if (random.nextFloat() < (modifier / (100.0f * (Math.floor((25.0F / f) + 1))))) { // Spigot - SPIGOT-7159: Better modifier resolution
                     // Spigot end
+                    boolean canGrow = // KioCG
                     CraftEventFactory.handleBlockGrowEvent(world, pos, this.getStateForAge(i + 1), 2); // CraftBukkit
+                    // KioCG start - 农业拓展
+                    if (canGrow) {
+                        BlockState belowState = world.getBlockState(pos.below());
+                        if (belowState.is(Blocks.FARMLAND)) {
+                            int moisture = (Integer) belowState.getValue(FarmBlock.MOISTURE);
+                            if (moisture > 0) CraftEventFactory.handleMoistureChangeEvent(world, pos.below(), belowState.setValue(FarmBlock.MOISTURE, moisture - 1), 2);
+                        }
+                    }
+                    // KioCG end - 农业拓展
                 }
             }
         }
diff --git a/src/main/java/net/minecraft/world/level/block/FarmBlock.java b/src/main/java/net/minecraft/world/level/block/FarmBlock.java
index 2b1bd583d9c7f049bfb798aa38ef021a881267e9..f269f4570f75c8c45062c5d3123ff00d4ba54189 100644
--- a/src/main/java/net/minecraft/world/level/block/FarmBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/FarmBlock.java
@@ -86,13 +86,7 @@ public class FarmBlock extends Block {
     public void randomTick(BlockState state, ServerLevel world, BlockPos pos, RandomSource random) {
         int i = (Integer) state.getValue(FarmBlock.MOISTURE);
 
-        if (!FarmBlock.isNearWater(world, pos) && !world.isRainingAt(pos.above())) {
-            if (i > 0) {
-                org.bukkit.craftbukkit.event.CraftEventFactory.handleMoistureChangeEvent(world, pos, (BlockState) state.setValue(FarmBlock.MOISTURE, i - 1), 2); // CraftBukkit
-            } else if (!FarmBlock.shouldMaintainFarmland(world, pos)) {
-                FarmBlock.turnToDirt((Entity) null, state, world, pos);
-            }
-        } else if (i < 7) {
+        if (i < 7 && world.isRainingAt(pos.above())) { // KioCG - 农业拓展
             org.bukkit.craftbukkit.event.CraftEventFactory.handleMoistureChangeEvent(world, pos, (BlockState) state.setValue(FarmBlock.MOISTURE, 7), 2); // CraftBukkit
         }
 
diff --git a/src/main/java/net/minecraft/world/level/block/StemBlock.java b/src/main/java/net/minecraft/world/level/block/StemBlock.java
index d95d5cd224b229fc34e271b56ca1dc9be13d6268..f59f2acb99879be3ffbda233f03189c896d91788 100644
--- a/src/main/java/net/minecraft/world/level/block/StemBlock.java
+++ b/src/main/java/net/minecraft/world/level/block/StemBlock.java
@@ -58,13 +58,30 @@ public class StemBlock extends BushBlock implements BonemealableBlock {
 
                 if (i < 7) {
                     state = (BlockState) state.setValue(StemBlock.AGE, i + 1);
+                    boolean canGrow = // KioCG
                     CraftEventFactory.handleBlockGrowEvent(world, pos, state, 2); // CraftBukkit
+                    // KioCG start - 农业拓展
+                    if (canGrow) {
+                        BlockState belowState = world.getBlockState(pos.below());
+                        if (belowState.is(Blocks.FARMLAND)) {
+                            int moisture = (Integer) belowState.getValue(FarmBlock.MOISTURE);
+                            if (moisture > 0) CraftEventFactory.handleMoistureChangeEvent(world, pos.below(), belowState.setValue(FarmBlock.MOISTURE, moisture - 1), 2);
+                        }
+                    }
+                    // KioCG end - 农业拓展
                 } else {
                     Direction enumdirection = Direction.Plane.HORIZONTAL.getRandomDirection(random);
                     BlockPos blockposition1 = pos.relative(enumdirection);
                     BlockState iblockdata1 = world.getBlockState(blockposition1.below());
 
                     if (world.getBlockState(blockposition1).isAir() && (iblockdata1.is(Blocks.FARMLAND) || iblockdata1.is(BlockTags.DIRT))) {
+                        // KioCG start - 农业拓展
+                        BlockState belowState = world.getBlockState(pos.below());
+                        if (!belowState.is(Blocks.FARMLAND)) return;
+                        int moisture = (Integer) belowState.getValue(FarmBlock.MOISTURE);
+                        if (moisture <= 0) return;
+                        CraftEventFactory.handleMoistureChangeEvent(world, pos.below(), belowState.setValue(FarmBlock.MOISTURE, moisture - 1), 2);
+                        // KioCG end - 农业拓展
                         // CraftBukkit start
                         if (!CraftEventFactory.handleBlockGrowEvent(world, blockposition1, this.fruit.defaultBlockState())) {
                             return;
