From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Joket_abc <15005771700@qq.com>
Date: Tue, 24 Oct 2023 15:42:36 +0800
Subject: [PATCH] Death drop protection


diff --git a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
index f33d6e9785acbd21a0b366587aa170cc1509bf58..68f3bad20b7977bf910f9e9d032c7753ff6f5b48 100644
--- a/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
+++ b/src/main/java/org/bukkit/craftbukkit/event/CraftEventFactory.java
@@ -1009,7 +1009,12 @@ public class CraftEventFactory {
         for (org.bukkit.inventory.ItemStack stack : event.getDrops()) {
             if (stack == null || stack.getType() == Material.AIR) continue;
 
+            // KioCG start
+            Item item =
             world.dropItem(entity.getLocation(), stack);
+            item.setOwner(entity.getUniqueId());
+            ((org.bukkit.craftbukkit.entity.CraftItem) item).getHandle().age = -32767;
+            // KioCG end
         }
 
         return event;
