From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Joket_abc <15005771700@qq.com>
Date: Mon, 13 Nov 2023 17:46:14 +0800
Subject: [PATCH] Fix openEnchanting method

https://github.com/PaperMC/Paper/issues/9943

diff --git a/src/main/java/org/bukkit/craftbukkit/entity/CraftHumanEntity.java b/src/main/java/org/bukkit/craftbukkit/entity/CraftHumanEntity.java
index 2a534391e49c0653fd98fe5c22e474ae7e82feef..2921edc7b0c3b344fd4a82e605fe1781730c0072 100644
--- a/src/main/java/org/bukkit/craftbukkit/entity/CraftHumanEntity.java
+++ b/src/main/java/org/bukkit/craftbukkit/entity/CraftHumanEntity.java
@@ -408,13 +408,14 @@ public class CraftHumanEntity extends CraftLivingEntity implements HumanEntity {
         // If there isn't an enchant table we can force create one, won't be very useful though.
         BlockPos pos = CraftLocation.toBlockPosition(location);
         // Paper start
-        MenuProvider menuProvider = ((EnchantmentTableBlock) Blocks.ENCHANTING_TABLE).getMenuProvider(null, this.getHandle().level(), pos);
+        final ServerLevel levelFinal = ((org.bukkit.craftbukkit.CraftWorld) location.getWorld()).getHandle(); // KioCG
+        MenuProvider menuProvider = ((EnchantmentTableBlock) Blocks.ENCHANTING_TABLE).getMenuProvider(null, levelFinal, pos); // KioCG - use the right world
         if (menuProvider == null) {
             if (!force) {
                 return null;
             }
             menuProvider = new net.minecraft.world.SimpleMenuProvider((syncId, inventory, player) -> {
-                return new net.minecraft.world.inventory.EnchantmentMenu(syncId, inventory, net.minecraft.world.inventory.ContainerLevelAccess.create(this.getHandle().level(), pos));
+                return new net.minecraft.world.inventory.EnchantmentMenu(syncId, inventory, net.minecraft.world.inventory.ContainerLevelAccess.create(levelFinal, pos)); // KioCG - use the right world
             }, Component.translatable("container.enchant"));
         }
         this.getHandle().openMenu(menuProvider);
