From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Joket_abc <15005771700@qq.com>
Date: Wed, 22 Nov 2023 23:42:24 +0800
Subject: [PATCH] Healing campfire


diff --git a/src/main/java/net/minecraft/world/entity/player/Player.java b/src/main/java/net/minecraft/world/entity/player/Player.java
index c023566a34b718b25938ff0abba4cca24e29469e..ac5fad8fd8df4e37321b9e26b75e63141eccd92d 100644
--- a/src/main/java/net/minecraft/world/entity/player/Player.java
+++ b/src/main/java/net/minecraft/world/entity/player/Player.java
@@ -184,6 +184,7 @@ public abstract class Player extends LivingEntity {
     public int sixRowEnderchestSlotCount = -1; // Purpur
     public int burpDelay = 0; // Purpur
     public boolean canPortalInstant = false; // Purpur
+    public int campfireHealTick; // KioCG
 
     // CraftBukkit start
     public boolean fauxSleeping;
diff --git a/src/main/java/net/minecraft/world/level/block/entity/CampfireBlockEntity.java b/src/main/java/net/minecraft/world/level/block/entity/CampfireBlockEntity.java
index f8247c12d91d3b2b25ca84f3e395846ea0c0bc6b..5c6724f5f6fa0ed68c40fdc44a39a67a0ab30bd2 100644
--- a/src/main/java/net/minecraft/world/level/block/entity/CampfireBlockEntity.java
+++ b/src/main/java/net/minecraft/world/level/block/entity/CampfireBlockEntity.java
@@ -119,12 +119,35 @@ public class CampfireBlockEntity extends BlockEntity implements Clearable {
             }
         }
 
+        // KioCG start
+        if (state.is(net.minecraft.world.level.block.Blocks.CAMPFIRE) && world.getGameTime() % 80L == 0L) {
+            CampfireBlockEntity.applyEffects(world, pos);
+        }
+        // KioCG end
+
         if (flag) {
             setChanged(world, pos, state);
         }
 
     }
 
+    // KioCG start
+    private static void applyEffects(Level world, BlockPos pos) {
+        net.minecraft.world.phys.AABB axisalignedbb = (new net.minecraft.world.phys.AABB(pos)).inflate(3, 2, 3).move(0.0, 1.0, 0.0);
+        java.util.List<net.minecraft.world.entity.player.Player> list1 = world.getEntitiesOfClass(net.minecraft.world.entity.player.Player.class, axisalignedbb);
+        if (!list1.isEmpty()) {
+            java.util.Iterator iterator = list1.iterator();
+            while (iterator.hasNext()) {
+                net.minecraft.world.entity.player.Player entityhuman = (net.minecraft.world.entity.player.Player) iterator.next();
+                if (entityhuman.campfireHealTick != entityhuman.tickCount) {
+                    entityhuman.campfireHealTick = entityhuman.tickCount;
+                    entityhuman.heal(1.0F, org.bukkit.event.entity.EntityRegainHealthEvent.RegainReason.REGEN);
+                }
+            }
+        }
+    }
+    // KioCG end
+
     public static void cooldownTick(Level world, BlockPos pos, BlockState state, CampfireBlockEntity campfire) {
         boolean flag = false;
 
