From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Jake Potrebic <jake.m.potrebic@gmail.com>
Date: Mon, 27 Mar 2023 10:20:00 -0700
Subject: [PATCH] Add Structure check API

https://github.com/PaperMC/Paper/pull/9062

diff --git a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
index a2062296bd6b8226c1b91f5b28b4a0cf6b48c73a..1b7fa9f9e7a45ecec2b7e4df9c486a1428314cba 100644
--- a/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
+++ b/src/main/java/org/bukkit/craftbukkit/CraftWorld.java
@@ -228,6 +228,12 @@ public class CraftWorld extends CraftRegionAccessor implements World {
         };
     }
     // Paper end
+    // Paper start - structure check API
+    @Override
+    public boolean hasStructureAt(final io.papermc.paper.math.Position position, final Structure structure) {
+        return this.world.structureManager().getStructureWithPieceAt(io.papermc.paper.util.MCUtil.toBlockPos(position), net.minecraft.resources.ResourceKey.create(net.minecraft.core.registries.Registries.STRUCTURE, CraftNamespacedKey.toMinecraft(structure.getKey()))).isValid();
+    }
+    // Paper end
 
     private static final Random rand = new Random();
 
