From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Joket_abc <15005771700@qq.com>
Date: Sun, 5 Nov 2023 05:28:47 +0800
Subject: [PATCH] Do not reset flight ability


diff --git a/src/main/java/net/minecraft/server/level/ServerPlayer.java b/src/main/java/net/minecraft/server/level/ServerPlayer.java
index 79b92aa192ffbec93c3e846e102f7fd8d5beefc3..f9cf764bb6b12783ad88b98da9ce6ecce566fbfb 100644
--- a/src/main/java/net/minecraft/server/level/ServerPlayer.java
+++ b/src/main/java/net/minecraft/server/level/ServerPlayer.java
@@ -1921,7 +1921,9 @@ public class ServerPlayer extends Player {
     public void restoreFrom(ServerPlayer oldPlayer, boolean alive) {
         this.wardenSpawnTracker = oldPlayer.wardenSpawnTracker;
         this.chatSession = oldPlayer.chatSession;
+        boolean oldMayFly = oldPlayer.getAbilities().mayfly; // KioCG
         this.gameMode.setGameModeForPlayer(oldPlayer.gameMode.getGameModeForPlayer(), oldPlayer.gameMode.getPreviousGameModeForPlayer());
+        this.getAbilities().mayfly = oldMayFly; // KioCG
         this.onUpdateAbilities();
         if (alive) {
             this.getInventory().replaceWith(oldPlayer.getInventory());
