From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Joket_abc <15005771700@qq.com>
Date: Sat, 24 Feb 2024 21:21:23 +0800
Subject: [PATCH] Transform spawned via mob spawner


diff --git a/src/main/java/net/minecraft/world/entity/Mob.java b/src/main/java/net/minecraft/world/entity/Mob.java
index c2c03fa2e221757bd48d0b8697412b3ad43ba9aa..f185340b56726dd84f22a6635db53cb19c8fc871 100644
--- a/src/main/java/net/minecraft/world/entity/Mob.java
+++ b/src/main/java/net/minecraft/world/entity/Mob.java
@@ -1543,6 +1543,7 @@ public abstract class Mob extends LivingEntity implements Targeting {
                     return null;
                 }
                 this.level().addFreshEntity(t0, spawnReason);
+                t0.spawnedViaMobSpawner = this.spawnedViaMobSpawner; // KioCG
                 // CraftBukkit end
                 if (this.isPassenger()) {
                     Entity entity = this.getVehicle();
diff --git a/src/main/java/net/minecraft/world/entity/animal/Cow.java b/src/main/java/net/minecraft/world/entity/animal/Cow.java
index be4ccc42d6f598cbaaf39aafbd49b594ac7b06fe..9d5904d807b33b392434d904820de5c7cde76117 100644
--- a/src/main/java/net/minecraft/world/entity/animal/Cow.java
+++ b/src/main/java/net/minecraft/world/entity/animal/Cow.java
@@ -233,6 +233,7 @@ public class Cow extends Animal {
             return InteractionResult.PASS;
         }
         this.level().addFreshEntity(mooshroom);
+        mooshroom.spawnedViaMobSpawner = this.spawnedViaMobSpawner; // KioCG
         this.remove(RemovalReason.DISCARDED, org.bukkit.event.entity.EntityRemoveEvent.Cause.DISCARD);
         if (!player.getAbilities().instabuild) {
             stack.shrink(1);
diff --git a/src/main/java/net/minecraft/world/entity/animal/MushroomCow.java b/src/main/java/net/minecraft/world/entity/animal/MushroomCow.java
index 21632120b52e4d594153ebe057a14afa74c3f4eb..31d7dc087ba1937218ab05617c15bb9fc17a6542 100644
--- a/src/main/java/net/minecraft/world/entity/animal/MushroomCow.java
+++ b/src/main/java/net/minecraft/world/entity/animal/MushroomCow.java
@@ -259,6 +259,7 @@ public class MushroomCow extends Cow implements Shearable, VariantHolder<Mushroo
                     return;
                 }
                 this.level().addFreshEntity(entitycow, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.SHEARED);
+                entitycow.spawnedViaMobSpawner = this.spawnedViaMobSpawner; // KioCG
 
                 this.discard(EntityRemoveEvent.Cause.TRANSFORMATION); // CraftBukkit - from above and add Bukkit remove cause
                 // CraftBukkit end
diff --git a/src/main/java/net/minecraft/world/entity/animal/frog/Tadpole.java b/src/main/java/net/minecraft/world/entity/animal/frog/Tadpole.java
index 2a9c2a69a0589e4e7b7c79d3716376b360a2eba1..ce133ba0ec78b2ffb0efd0b6cce7be1933b57b50 100644
--- a/src/main/java/net/minecraft/world/entity/animal/frog/Tadpole.java
+++ b/src/main/java/net/minecraft/world/entity/animal/frog/Tadpole.java
@@ -288,6 +288,7 @@ public class Tadpole extends AbstractFish {
                 // CraftBukkit end
                 this.playSound(SoundEvents.TADPOLE_GROW_UP, 0.15F, 1.0F);
                 worldserver.addFreshEntityWithPassengers(frog, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.METAMORPHOSIS); // CraftBukkit - add SpawnReason
+                frog.spawnedViaMobSpawner = this.spawnedViaMobSpawner; // KioCG
                 this.discard(EntityRemoveEvent.Cause.TRANSFORMATION); // CraftBukkit - add Bukkit remove cause
             }
         }
diff --git a/src/main/java/net/minecraft/world/entity/monster/Skeleton.java b/src/main/java/net/minecraft/world/entity/monster/Skeleton.java
index 5e6c102735326b383b8b72f4d230e98ee07cfd1a..647c06966a11c2f58f7f0964a7f0f4b9e34bea6e 100644
--- a/src/main/java/net/minecraft/world/entity/monster/Skeleton.java
+++ b/src/main/java/net/minecraft/world/entity/monster/Skeleton.java
@@ -234,6 +234,7 @@ public class Skeleton extends AbstractSkeleton {
         }
 
         this.level().addFreshEntity(skeleton);
+        skeleton.spawnedViaMobSpawner = this.spawnedViaMobSpawner; // KioCG
         this.remove(RemovalReason.DISCARDED, org.bukkit.event.entity.EntityRemoveEvent.Cause.DISCARD);
         if (!player.getAbilities().instabuild) {
             stack.shrink(1);
diff --git a/src/main/java/net/minecraft/world/entity/monster/Slime.java b/src/main/java/net/minecraft/world/entity/monster/Slime.java
index 0f77f00e9a02d1f982f285617604e7291b70a2a4..de3d08796aeb42fe7e13d834a9fca5ea135b74cc 100644
--- a/src/main/java/net/minecraft/world/entity/monster/Slime.java
+++ b/src/main/java/net/minecraft/world/entity/monster/Slime.java
@@ -345,6 +345,7 @@ public class Slime extends Mob implements Enemy {
             }
             for (LivingEntity living : slimes) {
                 this.level().addFreshEntity(living, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.SLIME_SPLIT); // CraftBukkit - SpawnReason
+                living.spawnedViaMobSpawner = this.spawnedViaMobSpawner; // KioCG
             }
             // CraftBukkit end
         }
diff --git a/src/main/java/net/minecraft/world/entity/npc/Villager.java b/src/main/java/net/minecraft/world/entity/npc/Villager.java
index 62ea4d030f5e387feb2852e1aa0c66041419763d..991d160c79b3017dfe5bc7ad826e202a1a84d78e 100644
--- a/src/main/java/net/minecraft/world/entity/npc/Villager.java
+++ b/src/main/java/net/minecraft/world/entity/npc/Villager.java
@@ -1027,6 +1027,7 @@ public class Villager extends AbstractVillager implements ReputationEventHandler
                     return;
                 }
                 world.addFreshEntityWithPassengers(entitywitch, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.LIGHTNING);
+                entitywitch.spawnedViaMobSpawner = this.spawnedViaMobSpawner; // KioCG
                 // CraftBukkit end
                 this.releaseAllPois();
                 this.discard(EntityRemoveEvent.Cause.TRANSFORMATION); // CraftBukkit - add Bukkit remove cause
