From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Joket_abc <15005771700@qq.com>
Date: Sun, 17 Dec 2023 14:26:29 +0800
Subject: [PATCH] Transform spawned via mob spawner


diff --git a/src/main/java/net/minecraft/world/entity/Mob.java b/src/main/java/net/minecraft/world/entity/Mob.java
index 6f42bec45e2c5d049ab0f3c83aa4a8c7d252529b..5a0a8ab8a63a7848ff6cdf81a23432fb354585a0 100644
--- a/src/main/java/net/minecraft/world/entity/Mob.java
+++ b/src/main/java/net/minecraft/world/entity/Mob.java
@@ -1535,6 +1535,7 @@ public abstract class Mob extends LivingEntity implements Targeting {
                     return null;
                 }
                 this.level().addFreshEntity(t0, spawnReason);
+                t0.spawnedViaMobSpawner = this.spawnedViaMobSpawner; // KioCG
                 // CraftBukkit end
                 if (this.isPassenger()) {
                     Entity entity = this.getVehicle();
diff --git a/src/main/java/net/minecraft/world/entity/animal/Cow.java b/src/main/java/net/minecraft/world/entity/animal/Cow.java
index 6d00b3cd4a9cb0fc8a9e9c27f37429a25fcf6e9d..f65a0dfe7b80c09f442b2ebcd210219dffdfbe6a 100644
--- a/src/main/java/net/minecraft/world/entity/animal/Cow.java
+++ b/src/main/java/net/minecraft/world/entity/animal/Cow.java
@@ -233,6 +233,7 @@ public class Cow extends Animal {
             return InteractionResult.PASS;
         }
         this.level().addFreshEntity(mooshroom);
+        mooshroom.spawnedViaMobSpawner = this.spawnedViaMobSpawner; // KioCG
         this.remove(RemovalReason.DISCARDED);
         if (!player.getAbilities().instabuild) {
             stack.shrink(1);
diff --git a/src/main/java/net/minecraft/world/entity/animal/MushroomCow.java b/src/main/java/net/minecraft/world/entity/animal/MushroomCow.java
index 3f00d40125b63a76ed549755d4c9d9a2ff2d4adf..c4800c716417716618880edc44609f898cd8235f 100644
--- a/src/main/java/net/minecraft/world/entity/animal/MushroomCow.java
+++ b/src/main/java/net/minecraft/world/entity/animal/MushroomCow.java
@@ -258,6 +258,7 @@ public class MushroomCow extends Cow implements Shearable, VariantHolder<Mushroo
                     return;
                 }
                 this.level().addFreshEntity(entitycow, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.SHEARED);
+                entitycow.spawnedViaMobSpawner = this.spawnedViaMobSpawner; // KioCG
 
                 this.discard(); // CraftBukkit - from above
                 // CraftBukkit end
diff --git a/src/main/java/net/minecraft/world/entity/animal/frog/Tadpole.java b/src/main/java/net/minecraft/world/entity/animal/frog/Tadpole.java
index 14135dac80dec4224715e21cef6556d22e569026..2a883c24548aa0208f9dfefb65b694d63ba62a4b 100644
--- a/src/main/java/net/minecraft/world/entity/animal/frog/Tadpole.java
+++ b/src/main/java/net/minecraft/world/entity/animal/frog/Tadpole.java
@@ -285,6 +285,7 @@ public class Tadpole extends AbstractFish {
                 // CraftBukkit end
                 this.playSound(SoundEvents.TADPOLE_GROW_UP, 0.15F, 1.0F);
                 worldserver.addFreshEntityWithPassengers(frog, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.METAMORPHOSIS); // CraftBukkit - add SpawnReason
+                frog.spawnedViaMobSpawner = this.spawnedViaMobSpawner; // KioCG
                 this.discard();
             }
         }
diff --git a/src/main/java/net/minecraft/world/entity/monster/Skeleton.java b/src/main/java/net/minecraft/world/entity/monster/Skeleton.java
index ede47e5da73097fec9435f9839646bf9df93336b..e5d2ddaf069c4498ebd08e1f8ea6bc8caaa97241 100644
--- a/src/main/java/net/minecraft/world/entity/monster/Skeleton.java
+++ b/src/main/java/net/minecraft/world/entity/monster/Skeleton.java
@@ -234,6 +234,7 @@ public class Skeleton extends AbstractSkeleton {
         }
 
         this.level().addFreshEntity(skeleton);
+        skeleton.spawnedViaMobSpawner = this.spawnedViaMobSpawner; // KioCG
         this.remove(RemovalReason.DISCARDED);
         if (!player.getAbilities().instabuild) {
             stack.shrink(1);
diff --git a/src/main/java/net/minecraft/world/entity/monster/Slime.java b/src/main/java/net/minecraft/world/entity/monster/Slime.java
index b18fc603f11cf946baf390b8621139c880a1979a..7f65f75c0fa3d69726e36617f5c270e1a4a541c9 100644
--- a/src/main/java/net/minecraft/world/entity/monster/Slime.java
+++ b/src/main/java/net/minecraft/world/entity/monster/Slime.java
@@ -337,6 +337,7 @@ public class Slime extends Mob implements Enemy {
             }
             for (LivingEntity living : slimes) {
                 this.level().addFreshEntity(living, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.SLIME_SPLIT); // CraftBukkit - SpawnReason
+                living.spawnedViaMobSpawner = this.spawnedViaMobSpawner; // KioCG
             }
             // CraftBukkit end
         }
diff --git a/src/main/java/net/minecraft/world/entity/npc/Villager.java b/src/main/java/net/minecraft/world/entity/npc/Villager.java
index 6b75e778c189780f14cbb33d9e84762671370c49..2ef2a881e42736d382580991bb32e5863ac0a39c 100644
--- a/src/main/java/net/minecraft/world/entity/npc/Villager.java
+++ b/src/main/java/net/minecraft/world/entity/npc/Villager.java
@@ -1025,6 +1025,7 @@ public class Villager extends AbstractVillager implements ReputationEventHandler
                     return;
                 }
                 world.addFreshEntityWithPassengers(entitywitch, org.bukkit.event.entity.CreatureSpawnEvent.SpawnReason.LIGHTNING);
+                entitywitch.spawnedViaMobSpawner = this.spawnedViaMobSpawner; // KioCG
                 // CraftBukkit end
                 this.releaseAllPois();
                 this.discard();
